### PR TITLE
Sorting state fixes

### LINKED ISSUE #572: States are not handled correctly when passed to non-stateful Seq2SeqEncoders

### PR DESCRIPTION
**Pull request by DeNeutoy**
There were 2 problems here:
- I forgot to sort and crop states passed to non-stateful encoders, so there were shape mis-matches.
- I hadn't wrapped zeros used to restore the correct shape to cropped states with `Variables`, so there were gradients being passed to tensors in the backward pass. 

I added 3 tests which all fail  in the first commit, fixed by the next one.
Fixes #572

### CODE REVIEW

File: tests/modules/encoder_base_test.py [hunk 1]
```diff
@@ -34,6 +34,26 @@ def setUp(self):
         self.sorting_indices = sorting_indices
         self.restoration_indices = restoration_indices
 
+    def test_non_stateful_states_are_sorted_correctly(self):
+        encoder_base = _EncoderBase(stateful=False)
+        initial_states = (Variable(torch.randn(6, 5, 7)),
```
Comment (matt-gardner) [inline tests/modules/encoder_base_test.py]: These shapes are confusing me - can you give a comment explaining the shape here?

Comment (matt-gardner) [reply tests/modules/encoder_base_test.py]: And this is a tuple because it's the hidden state and the memory cell for the LSTM?

File: tests/modules/encoder_base_test.py [hunk 2]
```diff
@@ -34,6 +34,26 @@ def setUp(self):
...
+        # Check that we sort the state for non-stateful encoders. To test
+        # we'll just use a "pass through" encoder, as we aren't actually testing
+        # the functionality of the encoder here anyway.
+        _, states, restoration_indices = encoder_base.sort_and_run_forward(lambda *x: x,
+                                                                           self.tensor,
+                                                                           self.mask,
+                                                                           initial_states)
+        zeros = torch.zeros([6, 2, 7])
+        for state, original in zip(states, initial_states):
+            assert list(state.size()) == [6, 3, 7]
```
Comment (matt-gardner) [inline tests/modules/encoder_base_test.py]: An explanation of what you're doing here would be helpful, as this is somewhat non-obvious.  Something like, "`sort_and_run_forward` strips fully-padded instances from the batch; in order to use the `restoration_indices` we need to add back the two that got stripped.  What we get back should match what we started with."

File: tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py [hunk 3]
```diff
@@ -124,6 +124,46 @@ def test_wrapper_raises_if_batch_first_is_false(self):
...
+        tensor[1, 4:, :] = 0
+        tensor[2, 0:, :] = 0
+        tensor[3, 6:, :] = 0
+        mask = torch.ones(5, 7)
+        mask[0, 3:] = 0
+        mask[1, 4:] = 0
+        mask[2, 0:] = 0
+        mask[3, 6:] = 0
+
+        initial_states = (Variable(torch.randn(6, 5, 7)),
```
Comment (matt-gardner) [inline tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py]: Oh, the 6 is because this is bidirectional?  Yeah, a shape comment would have helped a lot here.

File: tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py [hunk 4]
```diff
@@ -124,6 +124,46 @@ def test_wrapper_raises_if_batch_first_is_false(self):
             lstm = LSTM(bidirectional=True, num_layers=3, input_size=3, hidden_size=7)
             _ = PytorchSeq2SeqWrapper(lstm)
 
+    def test_wrapper_works_when_passed_state_with_zero_length_sequences(self):
+        lstm = LSTM(bidirectional=True, num_layers=3, input_size=3, hidden_size=7, batch_first=True)
+        encoder = PytorchSeq2SeqWrapper(lstm)
+        tensor = torch.rand([5, 7, 3])
+        tensor[0, 3:, :] = 0
```
Comment (matt-gardner) [inline tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py]: You don't need to zero out the tensor, just the mask, throughout the test.

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py [hunk 5]
```diff
@@ -87,6 +88,7 @@ def forward(self,  # pylint: disable=arguments-differ
         if num_valid < batch_size:
             _, length, output_dim = unpacked_sequence_tensor.size()
             zeros = unpacked_sequence_tensor.data.new(batch_size - num_valid, length, output_dim).fill_(0)
+            zeros = Variable(zeros)
```
Comment (matt-gardner) [inline allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: pytorch 0.4 will be nice...

File: tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py (hunk 3 above)
Comment (matt-gardner) [reply tests/modules/seq2seq_encoders/pytorch_seq2seq_wrapper_test.py]: You could also make one of these tests only use a single variable for the hidden state, instead of two, to make sure we cover that case correctly.

File: tests/modules/encoder_base_test.py (hunk 1 above)
Comment (DeNeutoy) [reply tests/modules/encoder_base_test.py]: oh yeah sorry that is weird - it's `num_layers * directions` - thanks pytorch 👍

File: allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py (hunk 5 above)
Comment (DeNeutoy) [reply allennlp/modules/seq2seq_encoders/pytorch_seq2seq_wrapper.py]: In 0.3, you can't concat variables with tensors, so this would have broken, which is still decent.
